tough/src/schema/mod.rs

```diff
@@ -324,6 +392,215 @@ impl Role for Targets {
+        let re = Regex::new(&regex_string);
+        if let Ok(re) = re {
+            re.is_match(&target)
+        } else {
+            false
+        }
+    }
+}
+
+impl Delegations {
+    ///Determines if target passes pathset specific matching
+    pub fn check_target(&self, target: &str) -> bool {
+        for role in &self.roles {
+            if role.paths.matched_target(target) {
+                return true;
+            }
+        }
+        false
+    }
+
+    ///Ensures that all delegated paths are allowed to be delegated
+    pub fn verify_paths(&self) -> Result<()> {
+        for sub_role in &self.roles {
+            for path in match &sub_role.paths {
+                PathSet::Paths(paths) => paths,
+                PathSet::PathHashPrefixes(paths) => paths,
+            } {
+                if !self.check_target(&path) {
+                    return Err(Error::UnmatchedPath {
+                        child: path.to_string(),
+                    });
+                }
+            }
+        }
+        Ok(())
+    }
+
+    ///Returns given role if its a child of struct
+    pub fn role(&self, role_name: &str) -> Option<&DelegatedRole> {
+        for role in &self.roles {
+            if role.name == role_name {
+                return Some(&role);
+            }
+        }
+        None
+    }
+
+    ///verifies that roles matches contain valid keys
+    pub fn verify_role(&self, role: &Signed<Targets>, name: &str) -> Result<()> {
+        let role_keys = self.role(name).expect("Role not found");
+        let mut valid = 0;
+
+        let mut data = Vec::new();
+        let mut ser = serde_json::Serializer::with_formatter(&mut data, CanonicalFormatter::new());
+        role.signed
+            .serialize(&mut ser)
+            .context(error::JsonSerialization {
+                what: format!("{} role", name.to_string()),
+            })?;
+
+        for signature in &role.signatures {
+            if role_keys.keyids.contains(&signature.keyid) {
+                if let Some(key) = self.keys.get(&signature.keyid) {
+                    if key.verify(&data, &signature.sig) {
+                        valid += 1;
+                    }
+                }
+            }
+        }
+
+        ensure!(
+            valid >= u64::from(role_keys.threshold),
+            error::SignatureThreshold {
+                role: RoleType::Targets,
+                threshold: role_keys.threshold,
+                valid,
+            }
+        );
+        Ok(())
+    }
+
+    ///Finds target using pre ordered search given `target_url` or error if the target is not found
+    pub fn find_target(&self, target_url: &str) -> Result<&Target> {
+        for del_role in &self.roles {
+            match &del_role.targets {
+                Some(targets) => match &targets.signed.find_target(target_url) {
+                    Ok(target) => return Ok(target),
+                    _ => continue,
+                },
+                None => continue,
```

Comment:
This can probably be simplified a bit and avoid some nested matching by doing:
```
if let Some(targets) = &delegated_role.targets {
...
}
```
You can then eliminate the `None => continue` and the loop will continue.